### PR TITLE
Added traceModel context

### PR DESCRIPTION
**Pull request by gertjanal**
Retrieve the default trace model or the model for a project, find property model by dotted name, like 'email.from'.

Added example;
`node examples/app.js` and browse to `localhost:3000/trace-model.html`

Additional fixes:
* Renamed `SessionManager.toJson` to `SessionManager.json` as it converts json to js object instead of "to json"
* Added keystore url to the examples

### CODE REVIEW

File: src/modules/traceModelContext.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { SessionManager } from './sessionManager.js';
```
Comment (gertjanal) [inline src/modules/traceModelContext.js:1]: Here it is

File: src/modules/traceModelContext.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    constructor(sessionManager, projectId) {
+        this.sessionManager = sessionManager;
+        this.url = projectId ? `/projects/${projectId}/tracemodel`: `/tracemodel`;
+    }
+
+    /**
+     * Get the trace model.
+     *
+     * @returns The trace model
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]:
```suggestion
     * @returns Promise for the trace model
```

File: src/modules/traceModelContext.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    get = () => {
+        if (!this.#traceModel) {
+            return this.sessionManager.gatekeeper(this.url).then(SessionManager.json).then((traceModel) => {
+                this.#traceModel = traceModel;
+                return traceModel;
+            });
+        }
+        return Promise.resolve(this.#traceModel);
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]: The request could be performed multiple times. Probably better to assign the request promise to `#traceModel`
```suggestion
        if (!this.#traceModel) {
            this.#traceModel = this.sessionManager.gatekeeper(this.url).then(SessionManager.json);
        }
        return this.#traceModel;
```

File: src/modules/traceModelContext.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                            return modelType.properties[split[2]];
+                        }
+                        if (split.length == 3) {
+                            // For example: email.misc.headerField
+                            return modelType.properties[split[1]];
+                        }
+                    }
+                }
+            }
+            return Promise.reject('Property ${property} not found in trace model');
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]:
```suggestion
            return Promise.reject(`Property ${property} not found in trace model`);
```

File: src/modules/traceModelContext.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        });
+    };
+
+    /**
+     * Get the model part describing the a trace type.
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]:
```suggestion
     * Get the model part describing a trace type.
```

File: src/modules/traceModelContext.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @param {string} type The name of the type, without origin or category.
+     * @returns The model of the type
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]:
```suggestion
     * @returns Promise for the model of the type
```

File: src/modules/traceModelContext.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            });
+        }
+        return Promise.resolve(this.#traceModel);
+    };
+
+    /**
+     * Get a trace model property by its dotted notation, as used in the queries.
+     *
+     * @param {string} property A dotted trace model property, e.g. `picture.width` or `email.misc.any`
+     * @returns An property model with description, type and unit
```
Comment (nfi-jd) [inline src/modules/traceModelContext.js]:
```suggestion
     * @returns Promise for the property model with description, type and unit
```

**Review by nfi-jd — CHANGES_REQUESTED** [review]

File: src/hansken-js.js [hunk 8]
```diff
@@ -2,10 +2,12 @@ import { ProjectContext } from './modules/projectContext.js';
 import { SinglefileContext } from './modules/singefileContext.js';
 import { Scheduler } from './modules/scheduler.js';
 import { SessionManager } from './modules/sessionManager.js';
+import { TraceModelContext } from './modules/traceModelContext.js';
 
 class HanskenClient {
 
     #scheduler;
+    #traceModelContexts = {};
```
Comment (ghost) [inline src/hansken-js.js]: maybe add a comment here the traceModelContexts is a map of projectId and trace models

File: src/modules/traceModelContext.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import { SessionManager } from './sessionManager.js';
+
+class TraceModelContext {
```
Comment (ghost) [inline src/modules/traceModelContext.js:3]: No class documentation?

File: src/modules/traceModelContext.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+    #traceModel;
+
+    /**
+     * Get the default trace model or a project trace model.
+     *
+     * @param {SessionManager} sessionManager The session manager, used for connections to the Hansken servers
+     * @param {UUID} projectId The projectId to get the traceModel from, or undefined when accessing the default trace model.
```
Comment (ghost) [inline src/modules/traceModelContext.js:11]:
```suggestion
     * @param {UUID} projectId The projectId to get the traceModel from, or don't provide one to use the default trace model.
```

File: src/modules/traceModelContext.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Get a trace model property by its dotted notation, as used in the queries.
+     *
+     * @param {string} property A dotted trace model property, e.g. `picture.width` or `email.misc.any`
+     * @returns An property model with description, type and unit
+     */
+    property = (property) => {
+        return this.get().then((model) => {
+            const split = property.split('\.');
+            if (split.length === 1) {
```
Comment (ghost) [inline src/modules/traceModelContext.js:64]: Can you add a comment what `split.length === 1` actually means? Does this mean we're dealing with intrinsic properties or something else?

**Review by ghost — APPROVED** [review]
Are we sure we can include our trace model as test code?
(Already heard it's okay, but leaving it in here as verification)

File: src/modules/traceModelContext.js (hunk 3 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Good idea

File: src/modules/traceModelContext.js (hunk 2 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Good idea

File: src/modules/traceModelContext.js (hunk 7 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Good idea

File: src/modules/traceModelContext.js (hunk 6 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Good idea

File: src/modules/traceModelContext.js (hunk 5 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Thnx

**Comment by gertjanal** [comment]
Verified, exposing the trace model to the public is no problem

File: src/hansken-js.js (hunk 8 above)
Comment (gertjanal) [reply src/hansken-js.js]: Added

File: src/modules/traceModelContext.js (hunk 10 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js:11]: thnx

File: src/modules/traceModelContext.js (hunk 11 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js:64]: ok

File: src/modules/traceModelContext.js (hunk 9 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js:3]: I think the documentation of the constructor is sufficient

File: src/modules/traceModelContext.js (hunk 4 above)
Comment (gertjanal) [reply src/modules/traceModelContext.js]: Well spotted

**Review by nfi-jd — APPROVED** [review]

File: examples/trace-model.html [hunk 12]
```diff
@@ -0,0 +1,26 @@
+<script type="module">
+    import {HanskenClient} from './src/hansken-js.js';
+    const hansken = new HanskenClient('/gatekeeper', '/keystore');
+
+    const properties = ['uid', 'privileged', 'does.not.exist', 'picture.width', 'email.misc.flag0x0392', 'data.raw.size', 'data.raw.hash.md5', '#note.modifiedOn'];
+
+    let findProperty;
```
Comment (LeoBlommers) [inline examples/trace-model.html:7]: can line 7 and 8 be combined. Just let findProperty = function( ...

File: src/hansken-js.js [hunk 13]
```diff
@@ -2,10 +2,12 @@ import { ProjectContext } from './modules/projectContext.js';
 import { SinglefileContext } from './modules/singefileContext.js';
 import { Scheduler } from './modules/scheduler.js';
 import { SessionManager } from './modules/sessionManager.js';
+import { TraceModelContext } from './modules/traceModelContext.js';
 
 class HanskenClient {
 
     #scheduler;
+    #traceModelContexts = {}; // {projectId, traceModelContext}
```
Comment (LeoBlommers) [inline src/hansken-js.js:10]: what does to comment add?

**Review by LeoBlommers — APPROVED** [review]

File: examples/trace-model.html (hunk 12 above)
Comment (gertjanal) [reply examples/trace-model.html:7]: the function invokes itself, so it needs to be declared first to fix this

File: src/hansken-js.js (hunk 13 above)
Comment (gertjanal) [reply src/hansken-js.js:10]: What the key and value are of the object

**Pull request merged**